extensions/azure/processors/PutAzureBlobStorage.cpp

```diff
@@ -0,0 +1,229 @@
+  context->getProperty(CommonStorageAccountEndpointSuffix, credentials.endpoint_suffix, flow_file);
+  context->getProperty(ConnectionString, credentials.connection_string, flow_file);
+  return credentials.getConnectionString();
+}
+
+void PutAzureBlobStorage::createAzureStorageClient(const std::string &connection_string, const std::string &container_name) {
+  // When used in multithreaded environment make sure to use the azure_storage_mutex_ to lock the wrapper so the
+  // client is not reset with different configuration while another thread is using it.
+  if (blob_storage_wrapper_ == nullptr) {
+    blob_storage_wrapper_ = minifi::utils::make_unique<storage::AzureBlobStorage>(connection_string, container_name);
+  } else {
+    blob_storage_wrapper_->resetClientIfNeeded(connection_string, container_name);
+  }
+}
+
+std::string PutAzureBlobStorage::getConnectionString(
+    const std::shared_ptr<core::ProcessContext> &context,
+    const std::shared_ptr<core::FlowFile> &flow_file) const {
+  auto connection_string = getAzureConnectionStringFromProperties(context, flow_file);
+  if (!connection_string.empty()) {
+    return connection_string;
+  }
+
+  return getConnectionStringFromControllerService(context);
+}
+
+void PutAzureBlobStorage::onTrigger(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSession> &session) {
+  logger_->log_debug("PutAzureBlobStorage onTrigger");
+  std::shared_ptr<core::FlowFile> flow_file = session->get();
+  if (!flow_file) {
+    return;
+  }
+
+  auto connection_string = getConnectionString(context, flow_file);
+  if (connection_string.empty()) {
+    logger_->log_error("Connection string is empty!");
+    session->transfer(flow_file, Failure);
+    return;
+  }
+
+  std::string container_name;
+  if (!context->getProperty(ContainerName, container_name, flow_file) || container_name.empty()) {
+    logger_->log_error("Container Name is invalid or empty!");
+    session->transfer(flow_file, Failure);
+    return;
+  }
+
+  std::string blob_name;
+  if (!context->getProperty(Blob, blob_name, flow_file) || blob_name.empty()) {
+    logger_->log_error("Blob name is invalid or empty!");
+    session->transfer(flow_file, Failure);
+    return;
+  }
+
+  utils::optional<azure::storage::UploadBlobResult> upload_result;
+  {
+    std::lock_guard<std::mutex> lock(azure_storage_mutex_);
+    createAzureStorageClient(connection_string, container_name);
+    if (create_container_) {
+      blob_storage_wrapper_->createContainer();
+    }
+    PutAzureBlobStorage::ReadCallback callback(flow_file->getSize(), *blob_storage_wrapper_, blob_name);
+    session->read(flow_file, &callback);
+    upload_result = callback.getResult();
+  }
+
+  if (!upload_result) {
+    logger_->log_error("Failed to upload blob '%s' to Azure storage container '%s'", blob_name, container_name);
+    session->transfer(flow_file, Failure);
+  } else {
```

Comment:
Minor, but adding the scope could be omitted if the unhappy path returned.

Reply:
Updated in [48ec7a](https://github.com/apache/nifi-minifi-cpp/pull/979/commits/48ec7a2ebb6d109c147f2a3b83aa2ae8996acfa2)